### PR TITLE
TINY-7700: Disabled menu items with children can still be opened

### PR DESCRIPTION
**Pull request by diogo405**
Related Ticket:

Description of Changes:
* Do not open disabled menu items

Pre-checks:
* [x] Changelog entry added
* [x] Tests have been added (if applicable)
* [x] Branch prefixed with `feature/` for new features (if applicable)

Review:
* [x] Milestone set
* [x] Review comments resolved

GitHub issues (if applicable):

### CODE REVIEW

**Review by diogo405 — COMMENTED** [review]
The diff doesn't help much so this is what I added to `expandRight`:
```
if (item.hasConfigured(Disabling) && Disabling.isDisabled(item)) {
      return Optional.none();
    } else {
...
```

File: modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledMenuItemTest.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+  const assertVerticalAlignMenuIsNotOpen = () => {
+    UiFinder.notExists(SugarBody.body(), '[role="menuitemcheckbox"]:contains("None")');
+    UiFinder.notExists(SugarBody.body(), '[role="menuitemcheckbox"]:contains("Top")');
+    UiFinder.notExists(SugarBody.body(), '[role="menuitemcheckbox"]:contains("Middle")');
+    UiFinder.notExists(SugarBody.body(), '[role="menuitemcheckbox"]:contains("Bottom")');
+  };
+
+  it('TINY-7700: Disabled menu item with children should not open on mouse hover', async () => {
+    const editor = await McEditor.pFromSettings(settings);
```
Comment (ashfordneil) [inline modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledMenuItemTest.ts:39]: Should this be done with `TinyHooks`? It looks like you're using the same editor config for all the tests

**Review by ashfordneil — APPROVED** [review]

File: modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledMenuItemTest.ts (hunk 1 above)
Comment (diogo405) [reply modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledMenuItemTest.ts:39]: My first go was using `TinyHooks`. I changed it cause for some reason (dunno) it was breaking `pSendKeysOn` from the last test. It was throwing an error about the selector `'[title="Vertical align"]'`. The only thing I can think of is that the previous editor was causing that. So I used the `McEditor` to add/remove the editor in each test and it worked. Dunno if that makes sense 🤷

File: modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledMenuItemTest.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    menu: {
+      table: { title: 'Table', items: 'inserttable | cell row column | tableclass tablecellclass tablecellvalign tablecellborderwidth tablecellborderstyle tablecaption tablecellbackgroundcolor tablecellbordercolor | advtablesort | tableprops deletetable' }
+    },
+    menubar: 'table',
+  };
+
+  const pOpenTableMenu = (editor) => {
+    const tableMenuItemSelector = '[role="menuitem"]:contains("Table")';
+    TinyUiActions.clickOnMenu(editor, tableMenuItemSelector);
+    return Waiter.pTryUntil('Wait for a specific menu to open', () => UiFinder.exists(SugarBody.body(), tableMenuItemSelector));
```
Comment (JamesToohey) [inline modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledMenuItemTest.ts:28]: This will always be present as you're using the same selector to open the menu.

That is, `tableMenuItemSelector` is always in the menubar. You need to use a selector to look for the vertical align button for this assertion instead.

`const verticalAlignMenuItemSelector = 'div[title="Vertical align"]';`

I suspect this is the root cause of your issues

File: modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledMenuItemTest.ts (hunk 1 above)
Comment (JamesToohey) [reply modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledMenuItemTest.ts:39]: We want to avoid instantiating editors whenever we can, so using `McEditor` sparingly. Often when using `TinyHooks` we need to tidy up between each test (closing dialogs, menus, clearing editor content). This can sometimes be done in the `beforeEach` and `afterEach` hooks.

See my other comments as to why your test wasn't working

File: modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledMenuItemTest.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    base_url: '/project/tinymce/js/tinymce',
+    plugins: 'table',
+    menu: {
+      table: { title: 'Table', items: 'inserttable | cell row column | tableclass tablecellclass tablecellvalign tablecellborderwidth tablecellborderstyle tablecaption tablecellbackgroundcolor tablecellbordercolor | advtablesort | tableprops deletetable' }
+    },
+    menubar: 'table',
+  };
+
+  const pOpenTableMenu = (editor) => {
+    const tableMenuItemSelector = '[role="menuitem"]:contains("Table")';
```
Comment (JamesToohey) [inline modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledMenuItemTest.ts:26]: You can move this selector out of this function

File: modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledMenuItemTest.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  const settings = {
+    base_url: '/project/tinymce/js/tinymce',
+    plugins: 'table',
+    menu: {
+      table: { title: 'Table', items: 'inserttable | cell row column | tableclass tablecellclass tablecellvalign tablecellborderwidth tablecellborderstyle tablecaption tablecellbackgroundcolor tablecellbordercolor | advtablesort | tableprops deletetable' }
+    },
+    menubar: 'table',
+  };
+
+  const pOpenTableMenu = (editor) => {
```
Comment (JamesToohey) [inline modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledMenuItemTest.ts:25]: Missing a type here 
```suggestion
  const pOpenTableMenu = (editor: Editor) => {
```

File: modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledMenuItemTest.ts (hunk 2 above)
Comment (JamesToohey) [reply modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledMenuItemTest.ts:28]: I'm also not sure this label makes sense? Should it be something like "Wait for table menu to open"?

**Review by JamesToohey — CHANGES_REQUESTED** [review]
Requesting changes so we can tidy up the `DisabledMenuItemTest`. This will likely mean cleaning up the open menus with a helper function between the tests.

- [x] Use TinyHooks for `DisabledMenuItemTest`
- [ ] Fix selectors in test

File: modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledMenuItemTest.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    UiFinder.notExists(SugarBody.body(), '[role="menuitemcheckbox"]:contains("Middle")');
+    UiFinder.notExists(SugarBody.body(), '[role="menuitemcheckbox"]:contains("Bottom")');
+  };
+
+  beforeEach(async () => {
+    await pOpenTableMenu();
+  });
+
+  afterEach(() => {
+    assertVerticalAlignMenuIsNotOpen();
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledMenuItemTest.ts:43]: Assertions should never be done in hooks, they should only contain setup and teardown logic.

- [ ] Revert this change please

File: modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledMenuItemTest.ts [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+import Theme from 'tinymce/themes/silver/Theme';
+
+describe('webdriver.tinymce.themes.silver.editor.menubar.DisabledMenuItemTest', () => {
+
+  const verticalAlignMenuItemSelector = '[title="Vertical align"]';
+  const tableMenuItemSelector = '[role="menuitem"]:contains("Table")';
+
+  const hook = TinyHooks.bddSetupLight<Editor>({
+    base_url: '/project/tinymce/js/tinymce',
+    plugins: 'table',
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledMenuItemTest.ts:17]: We generally try to avoid pulling in plugins in silver tests unless absolutely necessary. For this case it's not necessary, so can we please either move this test to the table plugin or just register our own nested menu and do assertions against that (preferably the latter option).

Comment (lnewson) [reply modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledMenuItemTest.ts:17]: To provide more context, it's not the table plugin logic we want to test from the theme perspective but instead the fact that a nested menu won't open when it's disabled. This can also help with maintainability, as if we refactor how the table plugin works then we have less places we have to make updates to (ie it's largely all isolated inside the table plugin).

File: modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledMenuItemTest.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
+import { Keys, Mouse, RealKeys, UiFinder, Waiter } from '@ephox/agar';
+import { afterEach, beforeEach, describe, it } from '@ephox/bedrock-client';
+import { TinyHooks, TinyUiActions } from '@ephox/mcagar';
+import { SugarBody } from '@ephox/sugar';
+
+import { Editor } from 'tinymce/core/api/PublicApi';
+import TablePlugin from 'tinymce/plugins/table/Plugin';
+import Theme from 'tinymce/themes/silver/Theme';
+
+describe('webdriver.tinymce.themes.silver.editor.menubar.DisabledMenuItemTest', () => {
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledMenuItemTest.ts:10]: Should this be called `DisabledNestedMenuItemTest`? As it's not testing a regular menu item, but instead a nested menu item.

File: modules/alloy/src/test/ts/browser/ui/dropdown/TieredMenuTest.ts [hunk 8]
```diff
@@ -37,18 +37,25 @@ UnitTest.asynctest('TieredMenuTest', (success, failure) => {
             items: Arr.map([
               { type: 'item', data: { value: 'a-alpha', meta: { text: 'a-Alpha' }}, hasSubmenu: false },
               { type: 'item', data: { value: 'a-beta', meta: { text: 'a-Beta' }}, hasSubmenu: true },
-              { type: 'item', data: { value: 'a-gamma', meta: { text: 'a-Gamma' }}, hasSubmenu: false }
+              { type: 'item', data: { value: 'a-gamma', meta: { text: 'a-Gamma', disabled: true }}, hasSubmenu: true }
```
[the remark this answers is not in the document]
Comment (lnewson) [reply modules/alloy/src/test/ts/browser/ui/dropdown/TieredMenuTest.ts:40]: This seems fine to me :shrug: This data structure is only used for tests anyways, so this seems as good of place as any given it doesn't really matter how it's passed in this instance.

File: modules/alloy/src/main/ts/ephox/alloy/ui/single/TieredMenuSpec.ts [hunk 9]
```diff
@@ -170,31 +171,35 @@ const make: SingleSketchFactory<TieredMenuDetail, TieredMenuSpec> = (detail, _ra
...
-      updateAriaExpansions(container, path);
-      // When expanding, always select the first.
-      return Optional.from(path[0]).bind((menuName) => layeredState.lookupMenu(menuName).bind((activeMenuPrep) => {
-        const activeMenu = buildIfRequired(container, menuName, activeMenuPrep);
-
-        // DUPE with above. Fix later.
-        if (!SugarBody.inBody(activeMenu.element)) {
-          Replacing.append(container, GuiFactory.premade(activeMenu));
-        }
+    if (item.hasConfigured(Disabling) && Disabling.isDisabled(item)) {
```
Comment (lnewson) [inline modules/alloy/src/main/ts/ephox/alloy/ui/single/TieredMenuSpec.ts:174]: This needs a changelog entry for Alloy as well.

File: modules/tinymce/CHANGELOG.md [hunk 10]
```diff
@@ -60,6 +60,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Using the Tab key to navigate into the editor on IE 11 would incorrectly focus the toolbar #TINY-3707
 - The editor selection could be placed in an incorrect location when undoing or redoing changes in a document containing `contenteditable="false"` elements #TINY-7663
 - Unbinding a native event handler inside the `remove` event caused an exception that blocked editor removal #TINY-7730
+- Disabled menu items could still be opened #TINY-7700
```
Comment (lnewson) [inline modules/tinymce/CHANGELOG.md:63]:
```suggestion
- Disabled nested menu items could still be opened #TINY-7700
```

File: modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledMenuItemTest.ts (hunk 6 above)
Comment (diogo405) [reply modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledMenuItemTest.ts:17]: Yep, makes sense 👍

File: modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledMenuItemTest.ts [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    plugins: 'table',
+    menu: {
+      table: { title: 'Table', items: 'inserttable | cell row column | tableclass tablecellclass tablecellvalign tablecellborderwidth tablecellborderstyle tablecaption tablecellbackgroundcolor tablecellbordercolor | advtablesort | tableprops deletetable' }
+    },
+    menubar: 'table',
+  }, [ Theme, TablePlugin ]);
+
+  const pOpenTableMenu = () => {
+    const editor = hook.editor();
+    TinyUiActions.keydown(editor, Keys.escape()); // to close table menu before opening
```
Comment (JamesToohey) [inline modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledMenuItemTest.ts:26]: Could we move this into an `afterEach` rather than it being in the helper function for opening the menu? I also think we should have `pOpenTableMenu` called in each test rather than in the `beforeEach`, otherwise it's difficult to discern what the tests are doing

Comment (diogo405) [reply modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledMenuItemTest.ts]: Done.

File: modules/alloy/CHANGELOG.md [hunk 12]
```diff
@@ -28,6 +28,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 - Fixed a number of positioning bugs when using an "inset" layout, especially any that renders at the bottom.
 - `LayoutInset` bubble classes were applied incorrectly, causing the bubble arrows to show on the opposite side.
 - `MakeshiftAnchorSpec` did not extend `HasLayoutAnchorSpec` which meant layouts couldn't be used for makeshift anchors.
+- Disabled nested menu items could still be opened.
```
Comment (lnewson) [inline modules/alloy/CHANGELOG.md:31]: This needs to be a little different to the TinyMCE changelog entry... so it probably needs to be something more like this:
```suggestion
- `TieredMenu` child menus were incorrectly opened when the parent menu item was disabled.
```

File: modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledNestedMenuItemTest.ts [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+        text: 'Preferences',
+        disabled: true,
+        getSubmenuItems: () => {
+          return [
+            {
+              type: 'menuitem',
+              text: 'Settings',
+            }
+          ];
+        }
```
Comment (lnewson) [inline modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledNestedMenuItemTest.ts:39]:
```suggestion
        getSubmenuItems: () => [{
          type: 'menuitem',
          text: 'Settings',
        }]
```
This can be slimmed down by using shorthand syntax.

**Review by lnewson — APPROVED** [review]

File: modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledNestedMenuItemTest.ts [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+    UiFinder.notExists(SugarBody.body(), '[role="menuitem"]:contains("Settings")');
+  };
+
+  afterEach(() => {
+    closeCodeMenu();
+  });
+
+  it('TINY-7700: Disabled menu item with children should not open on mouse hover', async () => {
+    await pOpenCodeMenu();
+    Mouse.hoverOn(SugarBody.body(), '[role="menuitem"]:contains("Preferences")');
```
Comment (JamesToohey) [inline modules/tinymce/src/themes/silver/test/ts/webdriver/editor/DisabledNestedMenuItemTest.ts:59]: Can we use `preferencesMenuItemSelector` here?

**Review by JamesToohey — APPROVED** [review]

**Review by ashfordneil — APPROVED** [review]
